### PR TITLE
Fix notification % escaping, debug UI

### PR DESCRIPTION
**Pull request by michaelkirk**
Also, consolidated the notifications debug UI (and fixed it) to make testing
this a bit easier.

PTAL @charlesmchen

### CODE REVIEW

File: Signal/src/ViewControllers/DebugUI/DebugUINotifications.swift [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+    override func section(thread aThread: TSThread?) -> OWSTableSection? {
+        guard let thread = aThread else {
+            owsFail("\(logTag) Notifications must specify thread.")
+            return nil
+        }
+
+        var sectionItems = [
+            OWSTableItem(title:"Last Incoming Message") {
+                Logger.info("\(self.logTag) scheduling notification for incoming message.")
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/DebugUI/DebugUINotifications.swift]: These `self` references are going to create retain cycles that will leak this UI.  I don't care about the debug UI, but we don't want to use this idiom in production code.  I wonder if there's a convenient way to do something safe in production code?

File: SignalMessaging/utils/DisplayableText.swift [hunk 2]
```diff
@@ -215,6 +215,17 @@ extension String {
         return text.ows_stripped()
     }
 
+    @objc
+    public class func filterNotificationText(_ text: String?) -> String? {
+        guard let text = self.filterText(text) else {
+            return nil
+        }
+
+        // Notifications strip anything that looks lik a printf formatting character,
```
Comment (charlesmchen) [inline SignalMessaging/utils/DisplayableText.swift]: s/lik/like/

Comment (charlesmchen) [reply SignalMessaging/utils/DisplayableText.swift]: s/printf formatting character/string formatting specifier/

File: Signal/src/environment/NotificationsManager.m [hunk 3]
```diff
@@ -312,9 +312,12 @@ - (void)notifyUserForIncomingMessage:(TSIncomingMessage *)message
                             [NSString stringWithFormat:NSLocalizedString(@"APN_MESSAGE_IN_GROUP_DETAILED", nil),
                                       senderName,
                                       threadName,
-                                      messageDescription];
+                                      [DisplayableText filterNotificationText:messageDescription]];
```
Comment (charlesmchen) [inline Signal/src/environment/NotificationsManager.m]: I don't understand why this is desirable.  The message description isn't the format string.  So formatting specifiers in it shouldn't be escaped.  If they are (as you have it here), won't they be rendered incorrectly?  E.g. won't "I'm with you 100%" be rendered as "I'm with you 100%%"?

File: Signal/src/environment/NotificationsManager.m [hunk 4]
```diff
@@ -312,9 +312,12 @@ - (void)notifyUserForIncomingMessage:(TSIncomingMessage *)message
...
+
                     } else {
-                        notification.alertBody = [NSString stringWithFormat:@"%@: %@", senderName, messageDescription];
+                        notification.alertBody = [NSString stringWithFormat:@"%@: %@",
+                                                           senderName,
+                                                           [DisplayableText filterNotificationText:messageDescription]];
```
Comment (charlesmchen) [inline Signal/src/environment/NotificationsManager.m]: Ditto.

File: Signal/src/environment/NotificationsManager.m (hunk 3 above)
Comment (michaelkirk) [reply Signal/src/environment/NotificationsManager.m]: Nope, that's the point of this change.

from https://developer.apple.com/documentation/uikit/uilocalnotification/1616646-alertbody

> Printf style escape characters are stripped from the string prior to display; to include a percent symbol (%) in the message, use two percent symbols (%%).

File: Signal/src/ViewControllers/DebugUI/DebugUINotifications.swift (hunk 1 above)
Comment (michaelkirk) [reply Signal/src/ViewControllers/DebugUI/DebugUINotifications.swift]: Thanks, fixed!

File: Signal/src/environment/NotificationsManager.m (hunk 3 above)
Comment (michaelkirk) [reply Signal/src/environment/NotificationsManager.m]: I've included some more details, and a link to the docs in the comment in DisplayableText

File: SignalMessaging/utils/DisplayableText.swift (hunk 2 above)
Comment (michaelkirk) [reply SignalMessaging/utils/DisplayableText.swift]: typo fixed - thanks. 

To be consistent with the iOS docs, I'm going to continue to refer to the affected chars as "printf formatting" chars.

**Comment by michaelkirk** [comment]
PTAL @charlesmchen

File: Signal/src/environment/NotificationsManager.m (hunk 3 above)
Comment (charlesmchen) [reply Signal/src/environment/NotificationsManager.m]: Oh. Can you please add a comment here (and maybe below too) with that link & quote?

Comment (michaelkirk) [reply Signal/src/environment/NotificationsManager.m]: Done!

**Pull request merged**